deltametrics/strat.py

```diff
@@ -32,3 +35,527 @@ def compute_compensation(line1, line2):
+        in the boxy stratigraphy. Columns in `strat_coords` correspond with
+        columns in `data_coords`.
+
+    data_coords : :obj:`ndarray`
+        A `3 x N` array of `t-x-y` coordinates where information is to be
+        extracted from the data array.  Columns in `data_coords` correspond
+        with columns in `strat_coords`.
+    """
+    # preallocate boxy arrays and helpers
+    plate = np.atleast_1d(np.zeros(strata.shape[1:], dtype=np.int8))
+    strat_coords, data_coords = [], []  # preallocate sparse idx lists
+    _zero = np.array([0])
+
+    # the main loop through the elevations
+    seek_elev = strata[-1, ...]  # the first seek is the last surface
+    for k in np.arange(len(z) - 1, -1, -1):  # for every z, from the top
+        e = z[k]  # which elevation for this iteration
+        whr = e < seek_elev  # where elev is below strat surface
+        t = np.maximum(_zero, (np.argmin(strata[:, ...] <= e, axis=0) - 1))
+        plate[whr] = int(1)  # track locations in the plate
+
+        xy = plate.nonzero()
+        ks = np.full((np.count_nonzero(plate)), k)  # might be faster way
+        idxs = t[xy]  # must happen before incrementing counter
+
+        strat_ks = np.column_stack((ks, *xy))
+        data_idxs = np.column_stack((idxs, *xy))
+        strat_coords.append(strat_ks)  # list of numpy arrays
+        data_coords.append(data_idxs)
+
+    strat_coords = np.vstack(strat_coords)
+    data_coords = np.vstack(data_coords)
+    return strat_coords, data_coords
+
+
+def _determine_strat_coordinates(elev, z=None, dz=None, nz=None):
+    """Return a valid Z array for stratigraphy based on inputs.
+
+    This helper function enables support for user specified `dz`, `nz`, or `z`
+    in many functions. The logic for determining how to handle these mutually
+    exclusive inputs is placed in this function, to ensure consistent behavior
+    across all classes/methods internally.
+
+    .. note::
+
+        At least one of the optional parameters must be supplied. Precedence
+        when multiple arguments are supplied is `z`, `dz`, `nz`.
+
+    Parameters
+    ----------
+    elev : :obj:`ndarray`
+        An up-to-three-dimensional array with timeseries of elevation values,
+        where elevation is expected to along the zeroth axis.
+
+    z : :obj:`ndarray`, optional
+        Array of Z values to use, returned unchanged if supplied.
+
+    dz : :obj:`float`, optional
+        Interval in created Z array. Z array is created as
+        ``np.arange(np.min(elev), np.max(elev)+dz, step=dz)``.
+
+    nz : :obj:`int`, optional
+        Number of intervals in `z`. Z array is created as 
+        ``np.linspace(np.min(elev), np.max(elev), num=nz, endpoint=True)``.
+    """
+    if (dz is None) and (z is None) and (nz is None):
+        raise ValueError('You must specify "z", "dz", or "nz.')
+
+    _valerr = ValueError('"dz" or "nz" cannot be zero or negative.')
+    if not (z is None):
+        if np.isscalar(z):
```

Comment:
I think this would be less general actually... `None` is the default for all three options, so we check explicitly if it is not `None` here. Regardless, you are probably right that there is a better way to handle these three mutually-exclusive options with precedence case.

I will revisit in #18